### PR TITLE
Adaptive step-size for discrete domains: the multivariate anisotropic case

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

We have non-adaptive rules and adaptives rules for the discrete case, but we have no adaptive rule that is coordinate-specific.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by teytaud** [comment]
@CarolaDoerr

File: nevergrad/parametrization/discretization.py [hunk 1]
```diff
@@ -42,9 +42,13 @@ def threshold_discretization(x: tp.ArrayLike, arity: int = 2) -> tp.List[int]:
 # The function below is the opposite of the function above.
 def inverse_threshold_discretization(indexes: tp.List[int], arity: int = 2) -> np.ndarray:
     indexes_arr = np.array(indexes, copy=True)
+    assert not np.any(np.isnan(indexes_arr))
     pdf_bin_size = 1 / arity
     # We take the center of each bin (in the pdf space)
-    return scipy.stats.norm.ppf(indexes_arr * pdf_bin_size + (pdf_bin_size / 2))  # type: ignore
+    x = scipy.stats.norm.ppf(indexes_arr * pdf_bin_size + (pdf_bin_size / 2))  # type: ignore
+    nan_indices = np.where(np.isnan(x))
+    x[nan_indices] = np.sign(indexes_arr[nan_indices]) * 500.
```
Comment (teytaud) [inline nevergrad/parametrization/discretization.py]: Let's avoid nan and use large values instead.

File: nevergrad/optimization/mutations.py [hunk 2]
```diff
@@ -69,7 +71,19 @@ def portfolio_discrete_mutation(self, parent: tp.ArrayLike, intensity: tp.Option
             boolean_vector = [self.random_state.rand() > (float(intensity) / dimension) for _ in parent]
         return [s if b else self.significantly_mutate(s, arity) for (b, s) in zip(boolean_vector, parent)]
 
+    def coordinatewise_mutation(self, parent: tp.ArrayLike, velocity: tp.ArrayLike, boolean_vector: tp.ArrayLike, arity: int) -> tp.ArrayLike:
+        """This is the anisotropic counterpart of the classical 1+1 mutations in discrete domains
+        with tunable intensity: it is useful for anisotropic adaptivity."""
+        dimension = len(parent)
+        boolean_vector = [False for _ in parent]
+        while not any(boolean_vector):
+            boolean_vector = [self.random_state.rand() < (1. / dimension) for _ in parent]
```
Comment (jrapin) [inline nevergrad/optimization/mutations.py:80]:
```suggestion
        boolean_vector = np.zeros(dimension, dtype=bool)
        while not boolean_vector.any():
            boolean_vector = self.random_state.rand(size=dimension) < (1. / dimension)
```

File: nevergrad/optimization/optimizerlib.py [hunk 3]
```diff
@@ -78,10 +80,16 @@ def __init__(
...
                 assert noise_handling[1] > 0.0, "the factor must be a float greater than 0"
                 assert noise_handling[0] in ["random", "optimistic"], f"Unkwnown noise handling: '{noise_handling}'"
-        assert mutation in ["gaussian", "cauchy", "discrete", "fastga", "doublefastga", "adaptive",
+        assert mutation in ["gaussian", "cauchy", "discrete", "fastga", "doublefastga", "adaptive", "coordinatewiseadaptive",
                             "portfolio", "discreteBSO", "lengler", "doerr"], f"Unkwnown mutation: '{mutation}'"
         if mutation == "adaptive":
             self._adaptive_mr = 0.5
+        if mutation == "coordinatewiseadaptive":
+            all_params = paramhelpers.flatten_parameter(self.parametrization)
+            arity = max(len(param.choices) if isinstance(param, p.TransitionChoice) else 500 for param in all_params.values())
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:89]: why is that duplicated from line 71? please remove if not needed

And I am not sure I like the 500 value, is there any reason for using this? this seems artificial

File: nevergrad/optimization/optimizerlib.py [hunk 4]
```diff
@@ -196,6 +207,12 @@ def _internal_tell(self, x: tp.ArrayLike, loss: tp.FloatLoss) -> None:
         if self.mutation == "adaptive":
             factor = 1.2 if loss <= self.current_bests["pessimistic"].mean else 0.731  # 0.731 = 1.2**(-np.exp(1)-1)
             self._adaptive_mr = min(1., factor * self._adaptive_mr)
+        if self.mutation == "coordinatewiseadaptive":
+            factor = 1.2 if loss < self.current_bests["pessimistic"].mean else 0.731  # 0.731 = 1.2**(-np.exp(1)-1)
+            for i in range(self.dimension):
+                if self._modified_variables[i]:
+                    self._velocity[i] *= factor 
+                    self._velocity[i] = min(max(self._velocity[i], 1.), self._arity / 4.)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:215]:
```suggestion
            inds = self._modified_variables
            self._velocity[inds] = np.clip(self._velocity[inds] * factor, 1., self._arity / 4.)
```
please verify, but that should be simpler and faster

File: nevergrad/parametrization/discretization.py [hunk 5]
```diff
@@ -42,9 +42,13 @@ def threshold_discretization(x: tp.ArrayLike, arity: int = 2) -> tp.List[int]:
...
 def inverse_threshold_discretization(indexes: tp.List[int], arity: int = 2) -> np.ndarray:
     indexes_arr = np.array(indexes, copy=True)
+    assert not np.any(np.isnan(indexes_arr))
     pdf_bin_size = 1 / arity
     # We take the center of each bin (in the pdf space)
-    return scipy.stats.norm.ppf(indexes_arr * pdf_bin_size + (pdf_bin_size / 2))  # type: ignore
+    x = scipy.stats.norm.ppf(indexes_arr * pdf_bin_size + (pdf_bin_size / 2))  # type: ignore
+    nan_indices = np.where(np.isnan(x))
+    x[nan_indices] = np.sign(indexes_arr[nan_indices] - (arity / 2.)) * 500.
+    return x
```
Comment (jrapin) [inline nevergrad/parametrization/discretization.py:51]: are those changes still needed if you are using 500 somewhere else? having such constants in 2 different place is dangereous.

**Review by jrapin — CHANGES_REQUESTED** [review]

File: nevergrad/optimization/mutations.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/optimization/mutations.py:80]: Thx.

File: nevergrad/optimization/optimizerlib.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py:215]: ah yes, better!

File: nevergrad/parametrization/discretization.py (hunk 5 above)
Comment (teytaud) [reply nevergrad/parametrization/discretization.py:51]: There is no reason for choosing 500 in both places. Actually in both places it should be something better.
In the other place, this should depend on other quantities in the problem.
Here, this is just something like a max float. Let me improve this.

File: nevergrad/optimization/optimizerlib.py (hunk 3 above)
Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py:89]: An educated guess could be like self.budget ** (1/3)    (inspired by progressive widening). However the fact that this is coordinate-wise makes the guess difficult.
This is "how much do we discretize in continuous domains".

Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py]: Duplicate code removed.

File: nevergrad/optimization/mutations.py [hunk 6]
```diff
@@ -69,7 +71,19 @@ def portfolio_discrete_mutation(self, parent: tp.ArrayLike, intensity: tp.Option
...
 
+    def coordinatewise_mutation(self, parent: tp.ArrayLike, velocity: tp.ArrayLike, boolean_vector: tp.ArrayLike, arity: int) -> tp.ArrayLike:
+        """This is the anisotropic counterpart of the classical 1+1 mutations in discrete domains
+        with tunable intensity: it is useful for anisotropic adaptivity."""
+        dimension = len(parent)
+        boolean_vector = np.zeros(dimension, dtype=bool)
+        while not any(boolean_vector):
+            boolean_vector = self.random_state.rand(dimension) < (1. / dimension)
+        discrete_data = discretization.threshold_discretization(parent, arity=arity)
+        return discretization.inverse_threshold_discretization([s if not b else s + np.random.choice([-1. ,1.]) * v for (b, s, v)
```
Comment (jrapin) [inline nevergrad/optimization/mutations.py:82]: can you please vectorize this?
something like:
discrete_data[boolean_vector] += np.random.choice([-1, 1], size=sum(boolean_vector))

**Review by jrapin — CHANGES_REQUESTED** [review]

File: nevergrad/optimization/mutations.py [hunk 7]
```diff
@@ -69,7 +71,25 @@ def portfolio_discrete_mutation(self, parent: tp.ArrayLike, intensity: tp.Option
...
+            boolean_vector = self.random_state.rand(dimension) < (1. / dimension)
+        discrete_data = discretization.threshold_discretization(parent, arity=arity)
+        discrete_data = np.where(
+            boolean_vector,
+            discrete_data + np.random.choice([-1., 1.], size=dimension) * velocity,
+            discrete_data)
+        return discretization.inverse_threshold_discretization(discrete_data)
+        #return discretization.inverse_threshold_discretization([s if not b else s + np.random.choice([-1. ,1.]) * v for (b, s, v) 
+        #                                                        in zip(boolean_vector, discrete_data, velocity)])
+
```
Comment (jrapin) [inline nevergrad/optimization/mutations.py:89]:
```suggestion
```

File: nevergrad/optimization/mutations.py [hunk 8]
```diff
@@ -69,7 +71,25 @@ def portfolio_discrete_mutation(self, parent: tp.ArrayLike, intensity: tp.Option
...
+        with tunable intensity: it is useful for anisotropic adaptivity."""
+        dimension = len(parent)
+        boolean_vector = np.zeros(dimension, dtype=bool)
+        while not any(boolean_vector):
+            boolean_vector = self.random_state.rand(dimension) < (1. / dimension)
+        discrete_data = discretization.threshold_discretization(parent, arity=arity)
+        discrete_data = np.where(
+            boolean_vector,
+            discrete_data + np.random.choice([-1., 1.], size=dimension) * velocity,
+            discrete_data)
```
Comment (jrapin) [inline nevergrad/optimization/mutations.py:85]: I was not aware of this option, nice :D

File: nevergrad/optimization/optimizerlib.py [hunk 9]
```diff
@@ -58,6 +58,8 @@ class _OnePlusOne(base.Optimizer):
     DoubleFastGA is an adaptation of FastGA to arity > 2, Portfolio corresponds to random mutation rates,
     DiscreteBSO corresponds to a decreasing schedule of mutation rate.
     Adaptive and Doerr correspond to various self-adaptive mutation rates.
+    
+    Coordinatewiseadaptive is the anisotropic counterpart of the adaptive version.
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:62]: case does not match with the actual name
maybe add some `_` for readibility, it's slightly long!

File: nevergrad/optimization/mutations.py (hunk 6 above)
Comment (teytaud) [reply nevergrad/optimization/mutations.py]: Done!

File: nevergrad/optimization/optimizerlib.py (hunk 9 above)
Comment (teytaud) [reply nevergrad/optimization/optimizerlib.py]: Done.

**Review by jrapin — APPROVED** [review]

**Pull request merged**